sdk/dotnet/Pulumi.Automation/Workspace.cs

```diff
@@ -259,8 +259,9 @@ public Task SelectStackAsync(string stackName)
         /// <param name="name">The name of the plugin.</param>
         /// <param name="version">The version of the plugin e.g. "v1.0.0".</param>
         /// <param name="kind">The kind of plugin e.g. "resource".</param>
+        /// <param name="options">Any additional plugin installation options.</param>
         /// <param name="cancellationToken">A cancellation token.</param>
-        public abstract Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, CancellationToken cancellationToken = default);
+        public abstract Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, PluginInstallOptions? options = null, CancellationToken cancellationToken = default);
```

Comment:
Technically, this is a binary breaking change. Though, I don't know if it'd affect anyone in practice since folks generally compile the source program each time, and it's unlikely anyone is passing a `cancellationToken` or overriding `InstallPluginAsync`.

If we did want to make it a binary compatible change, we'd need to add another overload. I can imagine two different approaches.

## Approach 1

Keep the existing method as-is, but change `abstract` to `virtual` and add a new `virtual` overload with the new parameter that throws:

```c#
public abstract class Workspace : IDisposable
{
    // Existing method changed from `abstract` to `virtual` and calls the new overload.
    public virtual Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, CancellationToken cancellationToken = default)
        => InstallPluginAsync(name, version, kind, null, cancellationToken);

    // New virtual overload, that just throws at runtime.
    public virtual Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, PluginInstallOptions? options = null, CancellationToken cancellationToken = default)
        => throw new NotSupportedException($"{nameof(InstallPluginAsync)} must be overridden");
}

public sealed class LocalWorkspace : Workspace
{
    public override Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, PluginInstallOptions? options = null, CancellationToken cancellationToken = default)
    {
        // ... current implementation in PR ...
    }
}
```

The upside of this approach is is only requires subclasses to override a single virtual method. The downside is that these are no longer abstract, so there's no way to force a subclass to implement it, aside from the runtime exception. Most of this is a moot point because in practice only our own `LocalWorkspace` will be subclassing it.

## Approach 2

Keep the existing methods as-is (remains `abstract`) and add a new virtual overload with the new option that calls the existing abstract overload, ignoring the passed-in `PluginInstallOptions`. It'd be up to a subclass to override either the existing abstract overload alone to support the current functionality, or both to be able to support the new `PluginInstallOptions`.

```c#
public abstract class Workspace : IDisposable
{
    // Existing abstract method with no changes.
    public abstract Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, CancellationToken cancellationToken = default);

    // New virtual method that can be overridden to support the new options.
    public virtual Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, PluginInstallOptions? options = null, CancellationToken cancellationToken = default)
    {
        // By default, call the existing abstract `InstallPluginAsync` overload, which ignores `PluginInstallOptions`.
        // A subclass can override both to add support for `PluginInstallOptions`.
        return InstallPluginAsync(name, version, kind, cancellationToken);
    }
}

public sealed class LocalWorkspace : Workspace
{
    public override Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, CancellationToken cancellationToken = default)
        => InstallPluginAsync(name, version, kind, null, cancellationToken);

    public override Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, PluginInstallOptions? options = null, CancellationToken cancellationToken = default)
    {
        // ... current implementation in PR ...
    }
}
```

If we do want to make a binary compatible change, is there a preference on which approach we'd want to take?

Reply:
```Though, I don't know if it'd affect anyone in practice since folks generally compile the source program each time, and it's unlikely anyone is passing a cancellationToken or overriding InstallPluginAsync.```

That's what I thought too, so it though we don't care? @mikhailshilkov for another opinion. Would help to know our stance on this so we can review PRs accordingly.

Reply:
`LocalWorkspace` has the same breaking change, right? Why do you think it's unlikely that somebody uses it in an incompatible way?

Reply:
Yea I was thinking about this and kind of regretting that I went with the optional `= default` for our `CancellationToken`s instead of doing separate overloads to begin with.

I'm up for either suggested approach, whichever you guys decide is best.

Reply:
Could we solve it by keeping the change as I made it, and then adding an extension to support the old overload?

Reply:
@orionstudt your suggestion LGTM, no binary compat I think but the old source code using the old form of the method will compile without changes. @justinvp @mikhailshilkov pinging is that allright? You both have written a lot more C# code than I have, I'm mostly just interested in moving the fix forward.

Reply:
Yeah, it's binary-breaking but I'd be fine to ship this with an explicit shout-out in the changelog.

Reply:
> but the old source code using the old form of the method will compile without changes

That's true for callers, but technically anyone can subclass `Workspace` and an override of `InstallPluginAsync` in a subclass won't be source compatible with the new signature. That said, I highly doubt anyone outside this assembly is subclassing `Workspace`. So I'm good with the binary break and breaking other subclasses of `Workspace`, as long as we maintain source compat for callers.

In terms of maintaining source compat, @orionstudt, is there any benefit to the extension method over keeping the old overload on the base class (no longer abstract with implementation that calls the new overload)?

```c#
public abstract class Workspace
{
    // Existing method, no longer abstract, calls new abstract overload.
    public Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, CancellationToken cancellationToken = default)
        => InstallPluginAsync(name, version, kind, null, cancellationToken);

    // New abstract overload.
    public abstract Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, PluginInstallOptions? options = null, CancellationToken cancellationToken = default);
}
```

Reply:
@justinvp Probably not, I actually would prefer the approach you just outlined because an extension would likely be in a separate file. Better that things remain close together. Assuming we are OK with potentially breaking the (highly unlikely) subclasses of `Workspace`?

Reply:
> Edit: I am getting this alert when trying to go with that approach.

Does the analyzer complain still when the defaults are removed from the old overload?

```c#
public abstract class Workspace
{
    public Task InstallPluginAsync(string name, string version, PluginKind kind, CancellationToken cancellationToken)
        => InstallPluginAsync(name, version, kind, null, cancellationToken);

    public abstract Task InstallPluginAsync(string name, string version, PluginKind kind = PluginKind.Resource, PluginInstallOptions? options = null, CancellationToken cancellationToken = default);
}
```

Reply:
Ok should be good now I just made the params on the back-compat overload non-optional.